repos/system_upgrade/el9toel10/actors/opensslenginescheck/libraries/opensslenginescheck.py

```diff
@@ -0,0 +1,154 @@
+# repos/system_upgrade/el8toel9/actors/opensslconfigcheck/libraries/opensslconfigcheck.py
+def _normalize_key(key):
+    """
+    Strip the part of the key before the first dot
+    """
+    s = key.split(".", 1)
+    if len(s) == 2:
+        return s[1]
+    return key
+
+
+def _key_equal(pair, key):
+    """
+    Check the keys are equal in OpenSSL configuration semantics
+
+    The OpenSSL semantics ignores everything before the first dot to allow specifying
+    something like following, where the first line would be otherwise normally ignored
+
+        TLS.MaxProtocol = TLSv1.3
+        DTLS.MaxProtocol = DTLSv1.2
+    """
+    if pair.key == key:
+        return True
+    return _normalize_key(pair.key) == key
+
+
+def _find_pair(block, key):
+    """
+    Find key-value pair in the given configuration block
+
+    In the given configuration block (OpenSslConfigBlock) find a key-value with a given key.
+    If multiple values match, only the last one is returned.
+    """
+    res = None
+    for pair in block.pairs:
+        if _key_equal(pair, key):
+            res = pair
+
+    return res
+
+
+def _openssl_find_block(config, name):
+    """
+    In the given configuration file (OpenSslConfig) find a block with a given name
+    """
+    for block in config.blocks:
+        if block.name == name:
+            return block
+
+    return None
+
+
+resources = [
+    reporting.RelatedResource('package', 'openssl'),
+    reporting.RelatedResource('file', '/etc/pki/tls/openssl.cnf')
+]
+
+
+def check_openssl_engines(config):
+    """
+    Check there are no engines configured in openssl.cnf
+
+    In case the of the pkcs11 engine, stop the upgrade as the package was removed from RHEL 10.
+
+    In case of other (usually third-party) engines, just warn about them being deprecated,
+    and suggest removal.
+    """
+    init_block = _openssl_find_block(config, config.openssl_conf)
+    if config.openssl_conf != "openssl_init" or not init_block:
+        reporting.create_report([
+            reporting.Title('Non-standard configuration of openssl.cnf'),
+            reporting.Summary(
+                'The OpenSSL configuration file `/etc/pki/tls/openssl.cnf` does not contain '
+                'expected initialization (openssl_conf = openssl_init key-value pair).'
+            ),
+            reporting.Remediation(
+                'The openssl.cnf file needs to contain the following initialization: '
+                '`openssl_conf = openssl_init` The `openssl_conf` now contains {} or '
+                'the `[ openssl_init ]` block is missing. '.format(config.openssl_conf)
+            ),
+            reporting.Groups([reporting.Groups.INHIBITOR]),
+            reporting.Severity(reporting.Severity.HIGH),
+            reporting.Groups([
+                    reporting.Groups.SECURITY,
+                    reporting.Groups.NETWORK,
+                    reporting.Groups.SERVICES
+            ]),
+        ] + resources)
+        return
```

Comment:
`openssl_conf` indicates the name of the specific section. It can be anything like `openssl_conf=blahblah`.

Ideally we should look for the `engines` section name from the section specified in `openssl_conf=some_section`.

Reply:
That is exactly what is done here. The `config.openssl_conf` has the name of the first first key=value pair.

Reply:
Oh. Sorry. LGTM then